File: orphaned-stock-rows-after-combination-delete/python/orphaned_stock_rows.py
"""Find and remove orphaned PrestaShop stock_available rows left behind
after a product combination is deleted.

There is no enforced cascade between combinations (product_attribute) and
stock_available, so deleting a combination through the Back Office or the
combinations webservice resource can leave its stock row behind. The Back
Office sums quantity across every stock_available row tied to a product, so
an orphan row with nonzero quantity silently inflates the displayed total
stock. This lists live combinations and all stock rows for a product, finds
rows whose id_product_attribute matches no live combination, and deletes
them only after re-confirming on a fresh fetch immediately beforehand.
Safe to run again and again.

Guide: https://www.allanninal.dev/prestashop/orphaned-stock-rows-after-combination-delete/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("orphaned_stock_rows")

BASE_URL = os.environ["PRESTASHOP_URL"].rstrip("/")
WS_KEY = os.environ["PRESTASHOP_WS_KEY"]
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
PRODUCT_IDS = [int(p) for p in os.environ.get("PRODUCT_IDS", "").split(",") if p.strip()]


def api_get(path, params):
    params = dict(params)
    params["output_format"] = "JSON"
    r = requests.get(f"{BASE_URL}/api/{path}", params=params, auth=(WS_KEY, ""), timeout=30)
    r.raise_for_status()
    return r.json()


def live_combinations(id_product):
    data = api_get("combinations", {"display": "full", "filter[id_product]": id_product})
    return data.get("combinations") or []


def stock_rows_for_product(id_product):
    data = api_get("stock_availables", {"display": "full", "filter[id_product]": id_product})
    rows = data.get("stock_availables") or []
    return [
        {
            "id": int(r["id"]),
            "id_product_attribute": int(r.get("id_product_attribute") or 0),
            "quantity": int(r.get("quantity") or 0),
            "out_of_stock": int(r.get("out_of_stock") or 0),
            "id_shop": int(r.get("id_shop") or 0),
        }
        for r in rows
    ]


def find_orphan_stock_rows(combinations, stock_rows):
    """Pure decision logic, no I/O.

    combinations: list of dicts from GET /api/combinations?filter[id_product]=X
                  (each with at least "id").
    stock_rows: list of dicts from GET /api/stock_availables?filter[id_product]=X
                (each with "id", "id_product_attribute", "quantity", "out_of_stock").

    Returns the stock rows whose id_product_attribute matches no live
    combination and is not 0 (0 is the base product's own stock row, which
    always survives regardless of combinations).
    """
    live_ids = {0} | {int(c["id"]) for c in combinations}
    return [row for row in stock_rows if int(row["id_product_attribute"]) not in live_ids]


def delete_stock_row(id_stock_available):
    r = requests.delete(
        f"{BASE_URL}/api/stock_availables/{id_stock_available}",
        auth=(WS_KEY, ""),
        timeout=30,
    )
    r.raise_for_status()


def run():
    total_orphan_quantity = 0
    removed = 0
    for id_product in PRODUCT_IDS:
        combinations = live_combinations(id_product)
        stock_rows = stock_rows_for_product(id_product)
        orphans = find_orphan_stock_rows(combinations, stock_rows)

        for orphan in orphans:
            total_orphan_quantity += orphan["quantity"]
            log.warning(
                "Product %s orphan stock row id=%s id_product_attribute=%s quantity=%s id_shop=%s (%s)",
                id_product, orphan["id"], orphan["id_product_attribute"],
                orphan["quantity"], orphan["id_shop"],
                "would delete" if DRY_RUN else "deleting",
            )
            if not DRY_RUN:
                # Re-fetch and re-diff right before deleting, to avoid a race
                # with a combination created between detection and repair.
                fresh_combinations = live_combinations(id_product)
                fresh_rows = stock_rows_for_product(id_product)
                still_orphan_ids = {o["id"] for o in find_orphan_stock_rows(fresh_combinations, fresh_rows)}
                if orphan["id"] in still_orphan_ids:
                    delete_stock_row(orphan["id"])
                    removed += 1
            else:
                removed += 1

    log.info(
        "Done. %d orphan row(s) %s, %d unit(s) of orphaned quantity found.",
        removed, "to delete" if DRY_RUN else "deleted", total_orphan_quantity,
    )


if __name__ == "__main__":
    run()

File: orphaned-stock-rows-after-combination-delete/node/orphaned-stock-rows.js
/**
 * Find and remove orphaned PrestaShop stock_available rows left behind
 * after a product combination is deleted.
 *
 * There is no enforced cascade between combinations (product_attribute) and
 * stock_available, so deleting a combination through the Back Office or the
 * combinations webservice resource can leave its stock row behind. The Back
 * Office sums quantity across every stock_available row tied to a product, so
 * an orphan row with nonzero quantity silently inflates the displayed total
 * stock. This lists live combinations and all stock rows for a product, finds
 * rows whose id_product_attribute matches no live combination, and deletes
 * them only after re-confirming on a fresh fetch immediately beforehand.
 * Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/prestashop/orphaned-stock-rows-after-combination-delete/
 */
import { pathToFileURL } from "node:url";

const BASE_URL = (process.env.PRESTASHOP_URL || "https://example.test").replace(/\/$/, "");
const WS_KEY = process.env.PRESTASHOP_WS_KEY || "dummy_key";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const PRODUCT_IDS = (process.env.PRODUCT_IDS || "").split(",").map((p) => p.trim()).filter(Boolean).map(Number);

function authHeader() {
  return "Basic " + Buffer.from(`${WS_KEY}:`).toString("base64");
}

async function apiGet(path, params) {
  const url = new URL(`${BASE_URL}/api/${path}`);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  url.searchParams.set("output_format", "JSON");
  const res = await fetch(url, { headers: { Authorization: authHeader() } });
  if (!res.ok) throw new Error(`PrestaShop ${res.status}`);
  return res.json();
}

async function liveCombinations(idProduct) {
  const data = await apiGet("combinations", { display: "full", "filter[id_product]": idProduct });
  return data.combinations || [];
}

async function stockRowsForProduct(idProduct) {
  const data = await apiGet("stock_availables", { display: "full", "filter[id_product]": idProduct });
  const rows = data.stock_availables || [];
  return rows.map((r) => ({
    id: Number(r.id),
    id_product_attribute: Number(r.id_product_attribute || 0),
    quantity: Number(r.quantity || 0),
    out_of_stock: Number(r.out_of_stock || 0),
    id_shop: Number(r.id_shop || 0),
  }));
}

/**
 * Pure decision logic, no I/O.
 *
 * combinations: array of objects from GET /api/combinations?filter[id_product]=X
 *               (each with at least "id").
 * stockRows: array of objects from GET /api/stock_availables?filter[id_product]=X
 *            (each with "id", "id_product_attribute", "quantity", "out_of_stock").
 *
 * Returns the stock rows whose id_product_attribute matches no live
 * combination and is not 0 (0 is the base product's own stock row, which
 * always survives regardless of combinations).
 */
export function findOrphanStockRows(combinations, stockRows) {
  const liveIds = new Set([0, ...combinations.map((c) => Number(c.id))]);
  return stockRows.filter((row) => !liveIds.has(Number(row.id_product_attribute)));
}

async function deleteStockRow(idStockAvailable) {
  const res = await fetch(`${BASE_URL}/api/stock_availables/${idStockAvailable}`, {
    method: "DELETE",
    headers: { Authorization: authHeader() },
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status}`);
}

export async function run() {
  let totalOrphanQuantity = 0;
  let removed = 0;

  for (const idProduct of PRODUCT_IDS) {
    const combinations = await liveCombinations(idProduct);
    const stockRows = await stockRowsForProduct(idProduct);
    const orphans = findOrphanStockRows(combinations, stockRows);

    for (const orphan of orphans) {
      totalOrphanQuantity += orphan.quantity;
      console.warn(
        `Product ${idProduct} orphan stock row id=${orphan.id} id_product_attribute=${orphan.id_product_attribute} quantity=${orphan.quantity} id_shop=${orphan.id_shop} (${DRY_RUN ? "would delete" : "deleting"})`
      );
      if (!DRY_RUN) {
        // Re-fetch and re-diff right before deleting, to avoid a race
        // with a combination created between detection and repair.
        const freshCombinations = await liveCombinations(idProduct);
        const freshRows = await stockRowsForProduct(idProduct);
        const stillOrphanIds = new Set(findOrphanStockRows(freshCombinations, freshRows).map((o) => o.id));
        if (stillOrphanIds.has(orphan.id)) {
          await deleteStockRow(orphan.id);
          removed++;
        }
      } else {
        removed++;
      }
    }
  }

  console.log(`Done. ${removed} orphan row(s) ${DRY_RUN ? "to delete" : "deleted"}, ${totalOrphanQuantity} unit(s) of orphaned quantity found.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
